snuba/query/processors/logical/filter_in_select_optimizer.py

```diff
@@ -0,0 +1,188 @@
+        if f.internal_function.function_name[-2:] == "If":
+            assert isinstance(f.parameters[1], FunctionCall)
+            return self._get_domain_of_predicate(f.parameters[1])
+        else:
+            return None
+
+    def _get_domain_of_predicate(self, predicate: FunctionCall) -> set[int] | None:
+        if predicate.function_name == "equals":
+            return self._get_domain_of_equals(predicate)
+        elif predicate.function_name == "in":
+            return self._get_domain_of_in(predicate)
+        elif predicate.function_name == "and":
+            return self._get_domain_of_and(predicate)
+        else:
+            return None
+
+    def _get_domain_of_equals(self, e: FunctionCall) -> set[int] | None:
+        assert e.function_name == "equals"
+        lhs = e.parameters[0]
+        if isinstance(lhs, Column) and lhs.column_name == "metric_id":
+            rhs = e.parameters[1]
+            assert isinstance(rhs, Literal)
+            assert isinstance(rhs.value, int)
+            return {rhs.value}
+        else:
+            return None
+
+    def _get_domain_of_in(self, e: FunctionCall) -> set[int] | None:
+        assert e.function_name == "in"
+        lhs = e.parameters[0]
+        rhs = e.parameters[1]
+        if isinstance(lhs, Column) and lhs.column_name == "metric_id":
+            assert isinstance(rhs, FunctionCall) and rhs.function_name == "array"
+            ids = set()
+            for metric_id in rhs.parameters:
+                assert isinstance(metric_id, Literal) and isinstance(
+                    metric_id.value, int
+                )
+                ids.add(metric_id.value)
+            return ids
+        else:
+            return None
+
+    def _get_domain_of_and(self, e: FunctionCall) -> Domain:
+        assert e.function_name == "and"
+        lhs = e.parameters[0]
+        rhs = e.parameters[1]
+        assert isinstance(lhs, FunctionCall) and isinstance(rhs, FunctionCall)
+        lres = self._is_mid_cond(lhs)
+        rres = self._is_mid_cond(rhs)
+        if lres == rres:
+            # unsupported, or no filtering
+            return None
+
+        # 'and' must have metric_id condition on one side, and irrelivance on the other
+        # in both of these cases the aggregate predicate is unsupported
+        if lres and self._contains_midcond(rhs):
+            return None
+        elif rres and self._contains_midcond(lhs):
+            return None
+
+        condside = lhs if lres else rhs
+        if condside.function_name == "equals":
+            return self._get_domain_of_equals(condside)
+        else:
+            return self._get_domain_of_in(condside)
+
+    def _domain_union(self, e1: Domain, e2: Domain) -> Domain:
+        """Given 2 domains, returns the union"""
+        if e1 is None or e2 is None:
+            return None
+        else:
+            return e1.union(e2)
+
+    def _is_mid_cond(self, f: FunctionCall) -> bool:
```

Comment:
not sure I understand this terminology

Reply:
This function would read better if it checked the positive instead of eliminating the negative.

`return f.function_name in ("equals", "in") and isinstance(f.parameters[0], Column) and f.parameters[0].column_name == "metric_id"`